File: VB/App_Code/MetadataArtsFileSystemProvider.vb
Imports System
Imports System.Collections.Generic
Imports System.Web.UI.WebControls
Imports System.Linq
Imports System.Data
Imports System.IO
Imports System.Web
Imports System.Web.UI
Imports DevExpress.Web

Public Class MetadataArtsFileSystemProvider
    Inherits ArtsFileSystemProvider


    Private allRecords_Renamed As List(Of ArtsFileSystemItem)
    Private ReadOnly Property AllRecords() As List(Of ArtsFileSystemItem)
        Get
            If allRecords_Renamed Is Nothing Then
                allRecords_Renamed = ArtsDataProvider.GetArts()
            End If
            Return allRecords_Renamed
        End Get
    End Property

    Public Sub New(ByVal rootFolder As String)
        MyBase.New(rootFolder)
    End Sub

    Public Overrides Iterator Function GetFiles(ByVal folder As FileManagerFolder) As IEnumerable(Of FileManagerFile)
        For Each file In MyBase.GetFiles(folder)
            Yield New FileManagerFile(Me, folder, file.FullName, New FileManagerFileProperties With {.Metadata = GetMetadataField(AllRecords, file)})
        Next file
    End Function

    Public Overrides Iterator Function GetFolders(ByVal parentFolder As FileManagerFolder) As IEnumerable(Of FileManagerFolder)
        For Each folder In MyBase.GetFolders(parentFolder)
            Yield New FileManagerFolder(Me, parentFolder, folder.FullName, New FileManagerFolderProperties With {.Metadata = GetMetadataField(AllRecords, folder)})
        Next folder
    End Function

    Private Function GetMetadataField(ByVal items As List(Of ArtsFileSystemItem), ByVal fileManagerItem As FileManagerItem) As Dictionary(Of String, Object)
        Dim result As New Dictionary(Of String, Object)()
        For Each item In items
            If item.ArtID.ToString() = fileManagerItem.Id Then
                result.Add("CreationDate", item.CreationDate.ToString().Trim())
                result.Add("OwnerUserName", item.OwnerUserName.ToString().Trim())
                result.Add("Type", item.Type.ToString().Trim())
                Return result
            End If
        Next item
        Return result
    End Function

End Class

File: CS/App_Code/MetadataArtsFileSystemProvider.cs
using System;
using System.Collections.Generic;
using System.Web.UI.WebControls;
using System.Linq;
using System.Data;
using System.IO;
using System.Web;
using System.Web.UI;
using DevExpress.Web;

public class MetadataArtsFileSystemProvider : ArtsFileSystemProvider
{
    private List<ArtsFileSystemItem> allRecords;
    private List<ArtsFileSystemItem> AllRecords {
        get {
            if (allRecords == null)
                allRecords = ArtsDataProvider.GetArts();
            return allRecords;
        }
    }

    public MetadataArtsFileSystemProvider(string rootFolder) : base(rootFolder) { }

    public override IEnumerable<FileManagerFile> GetFiles(FileManagerFolder folder)
    {
        foreach (var file in base.GetFiles(folder))
            yield return new FileManagerFile(this, folder, file.FullName,
                new FileManagerFileProperties
                {
                    Metadata = GetMetadataField(AllRecords, file),
                });
    }

    public override IEnumerable<FileManagerFolder> GetFolders(FileManagerFolder parentFolder)
    {
        foreach (var folder in base.GetFolders(parentFolder))
            yield return new FileManagerFolder(this, parentFolder, folder.FullName,
                new FileManagerFolderProperties
                {
                    Metadata = GetMetadataField(AllRecords, folder),
                });
    }

    private Dictionary<string, object> GetMetadataField(List<ArtsFileSystemItem> items, FileManagerItem fileManagerItem)
    {
        Dictionary<string, object> result = new Dictionary<string, object>();
        foreach (var item in items)
        {
            if(item.ArtID.ToString() == fileManagerItem.Id)
            {
                result.Add("CreationDate", item.CreationDate.ToString().Trim());
                result.Add("OwnerUserName", item.OwnerUserName.ToString().Trim());
                result.Add("Type", item.Type.ToString().Trim());
                return result;
            }
        }
        return result;
    }

}